src/mpid/ch4/shm/xpmem/xpmem_impl.h

```diff
@@ -8,5 +8,199 @@
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_recv(void *src_buf, MPI_Aint data_sz, void *recv_buf,
+                                                 MPL_atomic_int64_t * offset_ptr,
+                                                 intptr_t sreq_ptr, MPIR_Request * rreq)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    int fin_type;
+    mpi_errno = MPIDI_XPMEM_do_lmt_coop_copy(src_buf, data_sz, recv_buf, offset_ptr,
+                                             FALSE, &fin_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (fin_type == MPIDI_XPMEM_BOTH_FIN) {
+        /* tell sender we finished too */
+        int rank = MPIDIG_REQUEST(rreq, rank);
+        MPIR_Comm *comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(rreq, context_id));
+        mpi_errno = MPIDI_XPMEM_send_recv_fin(rank, comm, sreq_ptr);
+    }
+
+    if (fin_type == MPIDI_XPMEM_BOTH_FIN || fin_type == MPIDI_XPMEM_SENDER_FIN) {
+        /* sender is done, free the counter_ptr */
+        MPIR_Handle_obj_free(&MPIDI_XPMEM_cnt_mem, MPIDI_XPMEM_REQUEST(rreq, counter_ptr));
+    } else if (fin_type == MPIDI_XPMEM_RECVER_FIN) {
+        /* need wait for sender to tell me when to free the counter_ptr */
+        int c;
+        MPIR_cc_incr(&MPIDI_XPMEM_global.num_pending_cnt, &c);
+    }
+
+    if (fin_type != MPIDI_XPMEM_LOCAL_FIN) {
+        /* request is complete */
+        MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+        MPID_Request_complete(rreq);
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_send(void *src_buf, MPI_Aint data_sz, void *recv_buf,
+                                                 MPL_atomic_int64_t * offset_ptr,
+                                                 MPIR_Request * sreq, intptr_t rreq_ptr,
+                                                 int coop_counter_direct_flag,
+                                                 uint64_t coop_counter_offset)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    int fin_type;
+    mpi_errno = MPIDI_XPMEM_do_lmt_coop_copy(src_buf, data_sz, recv_buf, offset_ptr,
+                                             TRUE, &fin_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (fin_type == MPIDI_XPMEM_BOTH_FIN) {
+        /* tell recver we finished too */
+        int rank = MPIDIG_REQUEST(sreq, rank);
+        MPIR_Comm *comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(sreq, context_id));
+        mpi_errno = MPIDI_XPMEM_send_send_fin(rank, comm, rreq_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+    } else if (fin_type == MPIDI_XPMEM_RECVER_FIN) {
+        /* tell recver we finished so it can free the counter_ptr */
+        int rank = MPIDIG_REQUEST(sreq, rank);
+        MPIR_Comm *comm = MPIDIG_context_id_to_comm(MPIDIG_REQUEST(sreq, context_id));
+        mpi_errno = MPIDI_XPMEM_send_cnt_free(rank, comm,
+                                              coop_counter_direct_flag, coop_counter_offset);
+        MPIR_ERR_CHECK(mpi_errno);
```

Comment:
`MPIDI_XPMEM_send_cnt_free` is used in one place. I do not think we need subroutine for it.

Reply:
The purpose here is to group all the active message protocol code together. It is easier to read and more importantly easier to maintain. For example, if we decide to switch from internal control protocol to use just generic ch4-layer active message, we only need modify code there. Also in general, shorter code is easier to read, right?